wrappers/python/tests/model/test_llama_decoder.py

```diff
@@ -245,51 +214,17 @@ def test_forward_backward(self, starpu_simple, torch_rng,
         )
-        # MLP gradients
-        assert_close_by_frobnorm(
-            torch_layer.mlp.up_proj.weight.grad.detach().numpy(),
-            torch_layer_other.mlp.up_proj.weight.grad.detach().numpy(),
-            **dtype2tol[dtype]
-        )
-        assert_close_by_frobnorm(
-            torch_layer.mlp.gate_proj.weight.grad.detach().numpy(),
-            torch_layer_other.mlp.gate_proj.weight.grad.detach().numpy(),
-            **dtype2tol[dtype]
-        )
-        assert_close_by_frobnorm(
-            torch_layer.mlp.down_proj.weight.grad.detach().numpy(),
-            torch_layer_other.mlp.down_proj.weight.grad.detach().numpy(),
-            **dtype2tol[dtype]
-        )
-        # Normalizations gradients
-        assert_close_by_frobnorm(
-            torch_layer.post_attention_layernorm.weight.grad.detach().numpy(),
-            torch_layer_other.post_attention_layernorm.weight.grad.detach().numpy(),
-            **dtype2tol[dtype]
-        )
-        assert_close_by_frobnorm(
-            torch_layer.input_layernorm.weight.grad.detach().numpy(),
-            torch_layer_other.input_layernorm.weight.grad.detach().numpy(),
-            **dtype2tol[dtype]
-        )
-        # Attention gradients
-        assert_close_by_frobnorm(
-            torch_layer.self_attn.q_proj.weight.grad.detach().numpy(),
-            torch_layer_other.self_attn.q_proj.weight.grad.detach().numpy(),
-            **dtype2tol[dtype]
-        )
-        assert_close_by_frobnorm(
-            torch_layer.self_attn.v_proj.weight.grad.detach().numpy(),
-            torch_layer_other.self_attn.v_proj.weight.grad.detach().numpy(),
-            **dtype2tol[dtype]
-        )
-        assert_close_by_frobnorm(
-            torch_layer.self_attn.o_proj.weight.grad.detach().numpy(),
-            torch_layer_other.self_attn.o_proj.weight.grad.detach().numpy(),
-            **dtype2tol[dtype]
-        )
-        assert_close_by_frobnorm(
-            torch_layer.self_attn.k_proj.weight.grad.detach().numpy(),
-            torch_layer_other.self_attn.k_proj.weight.grad.detach().numpy(),
-            **dtype2tol[dtype]
-        )
+        for parameter_name, _ in torch_layer.named_parameters():
+            split_name = parameter_name.split(".")
+            if len(split_name) == 2:
+                assert_close_by_frobnorm(
+                    torch_layer.__getattr__(split_name[0]).weight.grad.detach().numpy(),
+                    torch_layer_other.__getattr__(split_name[0]).weight.grad.detach().numpy(),
+                    **dtype2tol[dtype]
+                )
+            elif len(split_name) == 3:
+                assert_close_by_frobnorm(
+                    torch_layer.__getattr__(split_name[0]).__getattr__(split_name[1]).weight.grad.detach().numpy(),
+                    torch_layer_other.__getattr__(split_name[0]).__getattr__(split_name[1]).weight.grad.detach().numpy(),
+                    **dtype2tol[dtype]
+                )
```

Comment:
You should never directly call `__getattr__`. Python has quite complex attribute lookup mechanism and `__getattr__` is a single piece of that machinery. Thus `getattr(obj, attr_name, default_value)` should be used instead.

In addition, if you need access multiple attributes like `x.y.z` then you can use `operator.attrgetter` as follows.
```python
from operator import attrgetter
x = ...  # Complex object.
z = attrgetter('y.z')(x)
```
You can likely meet it in sorting an iterable of complex objects.
```python
sorted(xs, key=attrgetter('y.z'))  # Use `x.y.z` as a sorting key.
```

Reply:
No `__getattr__` is used now. Instead, we iterate over `named_parameters()`.